### PR TITLE
[NEW] GraphQL API

### LINKED ISSUE #8221: GraphQL Support

### PR DESCRIPTION
**Pull request by kamilkisiela**
@RocketChat/core 

Closes #8221

YES! YES! YES!

GraphQL is finally here!

# Introduction

A couple of weeks ago we started to work on GraphQL implementation for Rocket.Chat.

For the server and also a corresponding GraphQL frontapp - [Angular PWA with apollo-client and GraphQL](https://github.com/RocketChat/Rocket.Chat.PWA/pull/25).

To make it work on top of the existing infrastructure we had to create three packages.

- [rocketchat:graphql](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-graphql)
- [rocketchat:accounts](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-accounts)
- [rocketchat:grant](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-grant)

I will explain each one of them separately.

We think there is a lot here that will help Meteor developers do the same transitions in their own apps.

This change will also help RocketChat scale better, making it easier for RocketChat to support different databases in the future, and to support multiple frontends and provide 3rd party API.

Let's start with something huge that has been around for quite some time and there's a lot of talking going on - GraphQL!

---

# GraphQL

## Server

The GraphQL server runs on [`express`](http://npmjs.com/package/express) thanks to great [`apollo-server-express`](http://npmjs.com/package/apollo-server-express) package.

## Managing schema

Since GraphQL contains schemas and resolvers. We have to match them together somehow.

We used [`graphql-tools`](http://npmjs.com/package/graphql-tools) because it allows to use the GraphQL schema language to build GraphQL.js schema, and also includes useful schema tools like per-type mocking. Thanks to it, implementing new things will be super easy.

To make everything well organized we modularized schemas and resolvers with [`merge-graphql-schemas`](http://npmjs.com/package/merge-graphql-schemas).

Each feature has it's own directory and contains Types, Queries, Mutations and others that belong to it separated into multiple files.

Even schemas are separated from resolvers. This way you can take a look at them conveniently and also share them easily with your dev team.

But there's more!

## Subscriptions

Any messaging app needs real-time, so we also implemented GraphQL [Subscriptions](https://dev-blog.apollodata.com/a-proposal-for-graphql-subscriptions-1d89b1934c18)! [The biggest feature](https://dev-blog.apollodata.com/the-next-step-for-realtime-data-in-graphql-b564b72eb07b) of GraphQL since the beginning!

It allows to listen to changes of data or custom events on the receiving side of an API, such as [Rocket.Chat.PWA](https://github.com/RocketChat/Rocket.Chat.PWA).

## Authentication

The authentication part.

We made it possible to use Rocket.Chat's and Meteor's existing accounts system in GraphQL thanks to the [`js-accounts`](http://github.com/js-accounts/accounts) stack - it's like Meteor's accounts packages but works with any server and frontend and can use GraphQL, REST or DDP under the hood.

## Common logic

We managed to reuse a lot of code.

But I think the most interesting is part of a code is the one that is responsible for creating new channels.

Inside GraphQL resolver:

https://github.com/kamilkisiela/Rocket.Chat/blob/9dd51b363d68e239185f9e0b90143dc135fe0427/packages/rocketchat-graphql/server/resolvers/channels/createChannel.js#L9-L30

Inside REST API endpoint:

https://github.com/kamilkisiela/Rocket.Chat/blob/9dd51b363d68e239185f9e0b90143dc135fe0427/packages/rocketchat-api/server/v1/channels.js#L178-L204

Common logic:

https://github.com/kamilkisiela/Rocket.Chat/blob/9dd51b363d68e239185f9e0b90143dc135fe0427/packages/rocketchat-api/server/v1/channels.js#L129-L169

Both, REST API and GraphQL API share the same logic, which we think is super cool.

**Used packages:**

- [@accounts/graphql-api](http://npmjs.com/package/@accounts/graphql-api)
- [apollo-server-express](http://npmjs.com/package/apollo-server-express)
- [express](http://npmjs.com/package/express)
- [graphql-subscriptions](http://npmjs.com/package/graphql-subscriptions)
- [graphql-tools](http://npmjs.com/package/graphql-tools)
- [merge-graphql-schemas](http://npmjs.com/package/merge-graphql-schemas)
- [subscriptions-transport-ws](http://npmjs.com/package/subscriptions-transport-ws)
- [rocketchat:accounts](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-accounts)
- [swydo:graphql](http://atmospherejs.com/swydo/graphql)

---

# Accounts

The only purpose of this package is to integrate the [`js-accounts`](https://github.com/js-accounts) stack with MongoDB and Meteor accounts system.

What is `js-accounts`? A fullstack authentication and accounts-management for GraphQL and REST.

By doing this we're able to easily use current RocketChat's users ecosystem with GraphQL.

**Used packages:**

- [@accounts/server](http://npmjs.com/package/@accounts/server)
- [@accounts/mongo](http://npmjs.com/package/@accounts/mongo)

---

# Grant

The main idea behind this package was to allow the external, stand-alone applications (i.e. [PWA](https://github.com/RocketChat/Rocket.Chat.PWA)) to be able to easily use OAuth authentication that is integrated with accounts system used in Rocket.Chat.

The flow:

1. User requests an authentication via some OAuth Provider. Let's use Google and PWA as an example.
2. An app redirects him to a URL handled by Grant. ( [_http://localhost:300/\_oauth\_apps/connect/google/pwa_](http://localhost:300/_oauth_apps/connect/google/pwa)).
3. The URL is being interpreted so Grant knows which provider should be used to authenticate a user and what is the app we want to redirect him after auth completes.
4. Authentication happens
5. OAuth Provider sends client back to Grant, according to predefined callback url.
6. Grant handles the response and redirects a user to the starting point with some data that allows the application to log user in.

What's the most important, every external application that has been registered in Grant  ecosystem shares the same OAuth credentials for each provider.
For example, imagine we have two PWA applications, React and Angular versions. We don't need to provide two sets of credentials for Google OAuth, we only use one and flow stays the same.

Beside registering providers, Grant allows to enable and disable them whenever you want to.
Imagine a situation where Rocket.Chat, the main Meteor application exposes only Google OAuth but we already defined Facebook and GitHub in Grant. You can just turn them off.

Grant also exposes an endpoint that shows which providers are enabled and used so you can make only those visible in your application.

We also created a few packages to have something to start with:

- [rocketchat:grant-**facebook**](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-grant-facebook)
- [rocketchat:grant-**google**](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-grant-google)
- [rocketchat:grant-**github**](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-grant-github)

It's super easy to add many more new providers. NPM's `grant` package [has a lot of built-in integrations](https://github.com/simov/grant#150-supported-providers--oauth-playground).

**Used packages:**

- [rocketchat:accounts](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-accounts)
- [express](http://npmjs.com/package/express)
- [express-session](http://npmjs.com/package/express-session)
- [grant-express](http://npmjs.com/package/grant-express)

---

# Everyone involved

I'm the author of the Pull Request but it doesn't mean I'm the only person behind it.

This is a group effort of a few friend developers who worked together:

@Urigo as a father of that project, I mean originator!  and the technical leader and reviewer of both the backend and the frontend code,

@tomermoshe and @eitanfr that developed PWA and helped with the schema.

@dotansimha and @davidyaha who gave technical guidance, help and supported my issues on their core libraries.

  

But of course that I actually wrote the code (and the [`apollo-angular`](http://npmjs.com/packages/apollo-angular) library)!!!

### CODE REVIEW

**Comment by marceloschmidt** [comment]
Regarding licensing and patent issues, let's wait until @leebyron has an update. Here's the conversation: https://github.com/facebook/graphql/issues/351

**Comment by kamilkisiela** [comment]
@marceloschmidt Spec available under OWFa 1.0

**Comment by gdelavald** [comment]
Hey, awesome job on this, I started doing some tests and will try to provide some feedback.

**Comment by gdelavald** [comment]
@kamilkisiela Hey, like I said above, amazing work on this, I've been testing the usual use cases we would need in an API and it's looking great. For now, a few problems I experienced:

1) Using /me and adding channels or directMessages in the return values seems to cause an issue related to Meteor async wrapper
 ```
Meteor code must always run within a Fiber.
Try wrapping callbacks that you pass to non-Meteor libraries with Meteor.bindEnvironment.
```

2) I might have missed this, but looks like there isn't a way to get another's user data, some use cases of this would be opening a DM (need a way to get the other user's id/username, `sendMessage` parameter `directTo` seem a little ambiguous here btw) or even opening another user's profile.

**Comment by kamilkisiela** [comment]
@gdelavald If you're running Meteor 1.6 it should be fixed, if older version then we could use rawCollection

**Comment by gdelavald** [comment]
@kamilkisiela I'll merge develop again so we can test with the updated Meteor.

Any thoughts on this?
> I might have missed this, but looks like there isn't a way to get another's user data, some use cases of this would be opening a DM (need a way to get the other user's id/username, sendMessage parameter directTo seem a little ambiguous here btw) or even opening another user's profile.

File: packages/rocketchat-graphql/server/helpers/authenticated.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { AccountsServer } from 'meteor/rocketchat:accounts';
+//import { authenticated as _authenticated } from '@accounts/graphql-api';
```
Comment (gdelavald) [inline packages/rocketchat-graphql/server/helpers/authenticated.js]: Are we going to use the mock or should we switch to the accounts one?

File: packages/rocketchat-graphql/server/api.js [hunk 2]
```diff
@@ -10,15 +10,25 @@ import cors from 'cors';
...
 import { executableSchema } from './schema';
 
-const subscriptionPort = 3100;
+const subscriptionPort = RocketChat.settings.get('Graphql_Subscription_Port') || 3100;
 
 // the Meteor GraphQL server is an Express server
 const graphQLServer = express();
 
-graphQLServer.use(cors());
+if (RocketChat.settings.get('Graphql_CORS')) {
```
Comment (graywolf336) [inline packages/rocketchat-graphql/server/api.js]: What happens if this setting is changed after the server has started? Does it require a restart of the server to have an effect?

File: packages/rocketchat-graphql/server/api.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import cors from 'cors';
+
+import { executableSchema } from './schema';
+
+const subscriptionPort = RocketChat.settings.get('Graphql_Subscription_Port') || 3100;
+
+// the Meteor GraphQL server is an Express server
+const graphQLServer = express();
+
+if (RocketChat.settings.get('Graphql_CORS')) {
```
Comment (graywolf336) [inline packages/rocketchat-graphql/server/api.js]: What happens if this setting is changed after the server has started? Does it require a restart of the server to have an effect?

Comment (gdelavald) [reply packages/rocketchat-graphql/server/api.js]: No, we added a middleware to check if it is enabled, this allow us to change the settings without a restart.

**Comment by bluetrade-invest** [comment]
Hey @kamilkisiela i don't can access links bellow. Can you help me? I have a lot problems with connection.

> To make it work on top of the existing infrastructure we had to create three packages.
> 
> * [rocketchat:graphql](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-graphql)
> * [rocketchat:accounts](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-accounts)
> * [rocketchat:grant](https://github.com/kamilkisiela/Rocket.Chat/tree/graphql/packages/rocketchat-grant)
>

**Comment by Urigo** [comment]
I believe GraphQL has been removed from RocketChat.

We would love to help bringing it back, a lot as changed and improved since 2 years ago

**Comment by rodrigok** [comment]
Thanks, @Urigo. You are right, GraphQL was removed, we are focusing right now to move everything possible to our REST API and then we will reevaluate the possibility to start a GraphQL API.

It would be awesome to receive contributions on a new spec for the new implementation and on toolings to test load and performance.

**Pull request merged**